src/main/java/uk/gov/ida/matchingserviceadapter/services/EidasMatchingRequestToLMSRequestTransform.java

```diff
@@ -0,0 +1,99 @@
+import uk.gov.ida.matchingserviceadapter.domain.MatchingServiceRequestContext;
+import uk.gov.ida.matchingserviceadapter.rest.MatchingServiceRequestDto;
+import uk.gov.ida.matchingserviceadapter.rest.matchingservice.EidasMatchingDataset;
+import uk.gov.ida.matchingserviceadapter.rest.matchingservice.LevelOfAssuranceDto;
+import uk.gov.ida.matchingserviceadapter.saml.UserIdHashFactory;
+import uk.gov.ida.saml.core.IdaConstants;
+import uk.gov.ida.saml.core.IdaConstants.Eidas_Attributes;
+import uk.gov.ida.saml.core.domain.AuthnContext;
+import uk.gov.ida.saml.core.extensions.eidas.BirthName;
+import uk.gov.ida.saml.core.extensions.eidas.CurrentFamilyName;
+import uk.gov.ida.saml.core.extensions.eidas.CurrentGivenName;
+import uk.gov.ida.saml.core.extensions.eidas.DateOfBirth;
+import uk.gov.ida.saml.core.extensions.eidas.Gender;
+import uk.gov.ida.saml.core.extensions.eidas.PersonIdentifier;
+import uk.gov.ida.saml.core.extensions.eidas.PlaceOfBirth;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static com.google.common.base.Optional.of;
+
+public class EidasMatchingRequestToLMSRequestTransform implements Function<MatchingServiceRequestContext, MatchingServiceRequestDto> {
+
+    private EidasToVerifyLoaTransformer eidasToVerifyLoaTransformer = new EidasToVerifyLoaTransformer();
+    private UserIdHashFactory userIdHashFactory;
+
+    public EidasMatchingRequestToLMSRequestTransform(UserIdHashFactory userIdHashFactory) {
+        this.userIdHashFactory = userIdHashFactory;
+    }
+
+    @Override
+    public MatchingServiceRequestDto apply(MatchingServiceRequestContext matchingServiceRequestContext) {
+        Assertion assertion = matchingServiceRequestContext.getAssertions().get(0);
+        List<Attribute> attributes = assertion.getAttributeStatements().get(0).getAttributes();
+
+        return new MatchingServiceRequestDto(extractEidasMatchingDataset(attributes),
+            com.google.common.base.Optional.absent(),
+            extractAndHashPid(assertion),
+            matchingServiceRequestContext.getAttributeQuery().getID(),
+            extractVerifyLoa(assertion));
+    }
+
+    private LevelOfAssuranceDto extractVerifyLoa(Assertion assertion) {
+        return eidasToVerifyLoaTransformer.apply(EidasLoa.valueOfUri(assertion.getAuthnStatements().get(0).getAuthnContext().getAuthnContextClassRef().getAuthnContextClassRef()));
+    }
+
+    private String extractAndHashPid(Assertion assertion) {
+        String euPid = assertion.getAttributeStatements()
+            .get(0)
+            .getAttributes()
+            .stream()
+            .filter(a -> a.getName().equals(IdaConstants.Eidas_Attributes.PersonIdentifier.NAME))
+            .findFirst()
+            .map(Attribute::getAttributeValues)
+            .map(vs -> vs.get(0))
+            .map(v -> ((PersonIdentifier) v).getPersonIdentifier())
+            .get();
```

Comment:
Is it just me finding this an excessive use of stream? (Especially inside of the lambda we have variables named a, vs and v?)

Reply:
I think my preference is not to keep chaining things to the point were we can tell what we are trying to do.